### PR TITLE
Reindex Throttle Checkpointing

### PR DESCRIPTION
**Pull request by henningandersen**
Rather than write checkpoints every time, we now only write at certain intervals.
Revamped the single-writer and finish handling around this to avoid any
waiting/blocking.

Relates #42612

### CODE REVIEW

**Comment by henningandersen** [comment]
ci/2 failure reported here: https://github.com/elastic/elasticsearch/issues/46907 so
@elasticmachine run elasticsearch-ci/2

**Review by Tim-Brooks — APPROVED** [review]
LGTM. I think we should also consider at some point just having a repeated scheduled checkpoint update at a certain interval. But this seems reasonable.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/ThrottlingConsumer.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    public void accept(T newValue) {
+        long now = nanoTimeSource.getAsLong();
+        synchronized (lock) {
+            if (closed) {
+                return;
+            }
+            this.value = newValue;
+            if (scheduledWrite == null) {
+                // schedule is non-blocking
+                scheduledWrite = threadPool.schedule(this::onScheduleTimeout, getDelay(now), ThreadPool.Names.SAME);
```
Comment (ywelsch) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/ThrottlingConsumer.java]: `schedule` loses the thread context. Given that we are afterwards writing to an index (possibly running with security enabled), I wonder if that thread context is relevant.

**Comment by henningandersen** [comment]
ci/1 failure to run `reindex/80_slices/Multiple slices with rethrottle` has been fixed here in #46967, intend to merge to FB with that failure.

**Pull request merged**